apps/user-office-backend/src/factory/zip/attachment.ts

```diff
@@ -0,0 +1,287 @@
+import { container } from 'tsyringe';
+
+import baseContext from '../../buildContext';
+import { Tokens } from '../../config/Tokens';
+import { GenericTemplateDataSource } from '../../datasources/GenericTemplateDataSource';
+import { ProposalDataSource } from '../../datasources/ProposalDataSource';
+import { QuestionaryDataSource } from '../../datasources/QuestionaryDataSource';
+import { SampleDataSource } from '../../datasources/SampleDataSource';
+import { TemplateDataSource } from '../../datasources/TemplateDataSource';
+import { DownloadOptions } from '../../middlewares/factory/factoryServices';
+import { GenericTemplate } from '../../models/GenericTemplate';
+import { Proposal } from '../../models/Proposal';
+import {
+  areDependenciesSatisfied,
+  getAllFields,
+  getTopicActiveAnswers,
+} from '../../models/ProposalModelFunctions';
+import { QuestionaryStep, Answer } from '../../models/Questionary';
+import { isRejection } from '../../models/Rejection';
+import { Sample } from '../../models/Sample';
+import { DataType, Question } from '../../models/Template';
+import { UserWithRole } from '../../models/User';
+import { GenericTemplatesArgs } from '../../resolvers/queries/GenericTemplatesQuery';
+import { QuestionsFilter } from '../../resolvers/queries/QuestionsQuery';
+import { SamplesArgs } from '../../resolvers/queries/SamplesQuery';
+import { getFileAttachments, Attachment } from '../util';
+
+export type ProposalAttachmentData = {
+  proposal: Proposal;
+  attachments: Attachment[];
+};
+
+const addProposalAttachments = (
+  proposalAttachmentData: ProposalAttachmentData,
+  questionarySteps: QuestionaryStep[],
+  questionIds: string[]
+) => {
+  const updatedProposalAttachmentData = { ...proposalAttachmentData };
+  for (const step of questionarySteps) {
+    if (!step) {
+      logger.logError('step not found', { ...questionarySteps }); // TODO: fix type of the second param in the lib (don't use Record<string, unknown>)
+
+      throw 'Could not download generated PDF';
+    }
+
+    const topic = step.topic;
+    const answers = getTopicActiveAnswers(questionarySteps, topic.id).filter(
+      (answer) =>
+        answer.question.dataType === DataType.FILE_UPLOAD &&
+        questionIds.includes(answer.question.id)
+    );
+
+    // if the questionary step has nothing else but `PROPOSAL_BASIS` question
+    // skip the whole step because the first page already has every related information
+    if (answers.length === 0) {
+      continue;
+    }
+
+    const questionaryAttachments: Attachment[] = [];
+
+    for (let i = 0; i < answers.length; i++) {
+      const answer = answers[i];
+
+      questionaryAttachments.push(...getFileAttachments(answer));
+    }
+    updatedProposalAttachmentData.attachments.push(...questionaryAttachments);
+  }
+
+  return updatedProposalAttachmentData;
+};
+const getQuestionIdsData = async (
+  getTemplates: (filter: {
+    questionIds: string[];
+  }) => Promise<Question[] | null>,
```

Comment:
Can we create type for the getTemplates somewhere else instead of jam everything inside? So that its more readable